### PR TITLE
refactor(versioning/loose): Strict null checks

### PR DESCRIPTION
**Pull request by zharinov**
## Changes:

- Fix type signatures

## Context:

- Ref: #12869

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/versioning/loose/generic.ts [hunk 1]
```diff
@@ -78,11 +78,19 @@ export const comparer = (
...
-    return versions.find((v) => equals(v, range)) || null;
-  }
-  function minSatisfyingVersion(versions: string[], range: string): string {
-    return versions.find((v) => equals(v, range)) || null;
+  function getSatisfyingVersion(
+    versions: string[],
+    range: string
+  ): string | null {
+    const result = versions.find((v) => equals(v, range));
+    return result?.length ? result : null;
```
Comment (viceice) [inline lib/versioning/loose/generic.ts:86]:
```suggestion
    return result ?? null;
```
Result of find is string or undefined.

So i think you only need to replace `||` with `??` in original code.

File: lib/versioning/loose/generic.ts [hunk 2]
```diff
@@ -78,11 +78,19 @@ export const comparer = (
...
+  }
+  function minSatisfyingVersion(
+    versions: string[],
+    range: string
+  ): string | null {
+    const result = versions.find((v) => equals(v, range));
+    return result?.length ? result : null;
```
Comment (viceice) [inline lib/versioning/loose/generic.ts:93]: See above

File: lib/versioning/loose/generic.ts [hunk 3]
```diff
@@ -19,26 +19,26 @@ export interface VersionComparator {
 // parse should return null if version not valid
 // parse should return an object with property release, an array of version sections major.minor.patch
 export const parser = (parse: VersionParser): Partial<VersioningApi> => {
-  function isValid(version: string): string {
+  function isValid(version: string): string | null {
```
Comment (viceice) [inline lib/versioning/loose/generic.ts:22]: This should return a boolean, can be done in separate PR.

File: lib/versioning/loose/generic.ts [hunk 4]
```diff
@@ -233,11 +241,13 @@ export abstract class GenericVersioningApi<
   }
 
   getSatisfyingVersion(versions: string[], range: string): string | null {
-    return versions.find((v) => this.equals(v, range)) || null;
+    const result = versions.find((v) => this.equals(v, range));
+    return result?.length ? result : null;
```
Comment (viceice) [inline lib/versioning/loose/generic.ts:245]: See other comments

File: lib/versioning/loose/generic.ts [hunk 5]
```diff
@@ -233,11 +241,13 @@ export abstract class GenericVersioningApi<
...
   }
 
   minSatisfyingVersion(versions: string[], range: string): string | null {
-    return versions.find((v) => this.equals(v, range)) || null;
+    const result = versions.find((v) => this.equals(v, range));
+    return result?.length ? result : null;
```
Comment (viceice) [inline lib/versioning/loose/generic.ts:250]: Same

File: lib/versioning/loose/utils.spec.ts [hunk 6]
```diff
@@ -83,8 +88,9 @@ describe('versioning/loose/utils', () => {
       ${'isVersion'}       | ${true}
       ${'matches'}         | ${true}
       ${'sortVersions'}    | ${0}
-    `('$fn', ({ fn, expected }) => {
-      expect(api[fn]()).toBe(expected);
+    `('$fn', ({ fn, expected }: ApiTestData) => {
+      const apiFn: () => ApiTestExpected = api[fn].bind(api) as never;
```
Comment (viceice) [inline lib/versioning/loose/utils.spec.ts:92]: 😕 looks pretty bad. We should split the test functions instead, otherwise we loose all type safety.

**Comment by viceice** [comment]
unblocked

File: lib/versioning/loose/utils.spec.ts [hunk 7]
```diff
@@ -34,8 +34,15 @@ describe('versioning/loose/utils', () => {
...
-        return _version === 'test' ? null : { release: [1, 0, 0] };
+      protected _parse(_version: string): GenericVersion | null {
+        const matchGroups = _version.match(
+          /^(?<major>\d)\.(?<minor>\d)\.(?<patch>\d)$/
+        )?.groups;
+        if (!matchGroups) {
+          return null;
+        }
+        const { major, minor, patch } = matchGroups;
+        return { release: [major, minor, patch].map(Number) };
```
Comment (viceice) [inline lib/versioning/loose/utils.spec.ts:45]:
```suggestion
        return { release: [major, minor, patch].map(n => parseInt(n, 10)) };
```
I think this is more safe

File: lib/versioning/loose/utils.spec.ts [hunk 8]
```diff
@@ -34,8 +34,15 @@ describe('versioning/loose/utils', () => {
...
-        return _version === 'test' ? null : { release: [1, 0, 0] };
+      protected _parse(_version: string): GenericVersion | null {
+        const matchGroups = _version.match(
+          /^(?<major>\d)\.(?<minor>\d)\.(?<patch>\d)$/
+        )?.groups;
+        if (!matchGroups) {
+          return null;
+        }
+        const { major, minor, patch } = matchGroups;
+        return { release: [major, minor, patch].map(n => parseInt(n, 10)) };
```
Comment (zharinov) [inline lib/versioning/loose/utils.spec.ts:45]:
```suggestion
        return { release: [major, minor, patch].map((n) => parseInt(n, 10)) };
```

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]

File: lib/versioning/loose/index.ts [hunk 9]
```diff
@@ -50,6 +50,10 @@ class LooseVersioningApi extends GenericVersioningApi {
         return part1 - part2;
       }
     }
+    // istanbul ignore if
+    if (!(parsed1.suffix && parsed2.suffix)) {
+      return 1;
+    }
```
Comment (zharinov) [inline lib/versioning/loose/index.ts:56]: @rarkins @viceice I think this is cause of Docker builds for swift

Comment (viceice) [reply lib/versioning/loose/index.ts:56]: No

**Pull request merged**